### PR TITLE
metadata: create specs secret mask

### LINKED ISSUE #24025: Add a job to generate spec_secrets_mask.yml
Add a job that is triggered on registry change that generates a spec_secrets_mask.yml file

### PR DESCRIPTION
**Pull request by alafanechere**
## What
Closes #24025
We want to generate the specs_secret_mask from the generated registry and the cached spec.

## How
* Catalog assets: create two new registries asset that are the registries built from metadata to which we add the spec value from the cached spec downloaded from GCS:
`cloud_catalog_from_metadata_and_spec`
`oss_catalog_from_metadata_and_spec`

* Spec secret masks assets:
*  `all_specs_secrets`: A list of all the secret fields built by scanning all the `airbyte_secret` fields of the spec found int the registries.
*  `specs_secrets_mask_yaml`: the yaml representation of all_specs_secrets

Here's an example of the generated yaml:
```yaml
properties:
- application_key
- token_secret
- databricks_personal_access_token
- service_account
- api_token
- appid
- aws_key_secret
- aws_key_id
- access_key_id
- access_key
- access_key_secret
- application_secret
- org_id
- app_secret_token
- aws_secret_key
- tenant_id
- X-Postmark-Server-Token
- realm_id
- apiKeySecret
- app_api_key
- auth_user_password
- session_token
- api_key
- site_api_key
- agreement_grant_token
- hmac_key_secret
- client_secret_value
- site_id
- jwt_token
- api_password
- user_secret
- tunnel_user_password
- auth_hash
- client_certificate
- survey_id
- azure_blob_storage_sas_token
- shared_key
- s3_access_key_id
- user_token
- api_secret
- private_key
- app_id
- sasl_jaas_config
- user_name
- client_secret
- storage_access_key
- aws_access_key_id
- aws_secret_access_key
- api_key_secret
- ca_certificate
- access_token
- public_key
- key_store_password
- role_arn
- developer_token
- consumer_secret
- account_sid
- accessKey
- ssl_ca_certificate
- secret_key
- hmac_key_access_id
- password
- privateKey
- credentials_json
- apiId
- personal_access_token
- aws_access_key
- apikey
- auth_ssh_key
- restid
- client_key_password
- client_key
- ssh_key
- s3_secret_access_key
- token_uri
- sas_token
- service_account_info
- google_application_credentials
- client_id
- survey_group_id
- apiKey
- X-Postmark-Account-Token
- secret_access_key
- key
- ssl_certificate
- azure_blob_storage_account_key
- secret
- private_key_password
- consumer_key
- token_key
- token
- lwa_client_secret
- account_id
- refresh_token
- key_encrypting_key
- user_auth_key
- realm
- domain_url
- reserved_attribute_names
- auth_token
```

### CODE REVIEW

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/spec_cache.py [hunk 1]
```diff
@@ -56,3 +61,9 @@ def list_cached_specs() -> List[CachedSpec]:
     blobs = bucket.list_blobs(prefix=CACHE_FOLDER)
 
     return [get_docker_info_from_spec_cache_path(blob.name) for blob in blobs]
+
+
+def get_cached_spec(spec_cache_path: str) -> dict:
+    client = storage.Client.create_anonymous_client()
+    bucket = client.bucket(SPEC_CACHE_BUCKET_NAME)
+    return json.loads(bucket.blob(spec_cache_path).download_as_string())
```
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/lib/metadata_service/spec_cache.py:69]: @bnchrch should we download specs from the metadata_service lib or in the Dagster realm?

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/__init__.py [hunk 2]
```diff
@@ -82,6 +84,7 @@
     ),
     "latest_oss_catalog_gcs_file": gcs_file_blob.configured({"gcs_prefix": CATALOG_FOLDER, "gcs_filename": "oss_catalog.json"}),
     "latest_cloud_catalog_gcs_file": gcs_file_blob.configured({"gcs_prefix": CATALOG_FOLDER, "gcs_filename": "cloud_catalog.json"}),
+    "specs_secrets_mask": gcs_file_blob.configured({"gcs_prefix": CATALOG_FOLDER, "gcs_filename": "specs_secrets_mask.yaml"}),
```
Comment (alafanechere) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/__init__.py:87]: @bnchrch I did not upload the specs_secrets_mask to GCS. I'd love a inputs from your side how you want to orchestrate this kind of operation. We're not yet uploading registries to GCS so I did not find a example to copy 😄

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/spec_cache.py (hunk 1 above)
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/spec_cache.py:69]: Honestly, Im neutral on this one.

To me gcs blobs are used enough to be considered a "type" that we could build a contract between two systems on.

So if the metadata service returns a list of gcs blobs for the metadata files and its up to the calling service if they download them or something else. Thats fine in my books. 

But its also fine if the service returns the raw text from the file.

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/__init__.py (hunk 2 above)
Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/__init__.py:87]: There is no story currently! But lets do it in this PR!

Heres an example of how we already write to the reports/ folder in the GCS bucket

https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog_report.py#L125

Comment (bnchrch) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/__init__.py:87]: Also I think we dont have to have "specs_secrets_mask" as a resource here as were currently not triggering anything from the file itself at the moment

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/specs_secrets_mask.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        + oss_catalog_from_metadata_and_spec["destinations"]
+        + cloud_catalog_from_metadata_and_spec["destinations"]
+    )
+    for registry_entry in all_entries:
+        all_secret_properties += get_secrets_properties_from_registry_entry(registry_entry)
+    return set(all_secret_properties)
+
+
+@asset(group_name=GROUP_NAME)
+def specs_secrets_mask_yaml(all_specs_secrets: Set[str]) -> str:
```
Comment (bnchrch) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/specs_secrets_mask.py:51]: 👍

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py [hunk 4]
```diff
@@ -139,6 +171,16 @@ def oss_catalog_from_metadata(catalog_derived_metadata_definitions: List[Partial
     return construct_catalog_from_metadata(catalog_derived_metadata_definitions, "oss")
 
 
+@asset(group_name=GROUP_NAME)
+def cloud_catalog_from_metadata_and_spec(cloud_catalog_from_metadata: dict, cached_specs: OutputDataFrame) -> dict:
```
Comment (bnchrch) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/catalog.py:175]: I'm thinking we should merge `*_catalog_from_metadata_and_spec` into `*_catalog_from_metadata`

File: airbyte-ci/connectors/metadata_service/orchestrator/tests/test_specs_secrets_mask.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+import pytest
+from orchestrator.assets.specs_secrets_mask import get_secrets_properties_from_registry_entry
+
+
+@pytest.mark.parametrize(
```
Comment (bnchrch) [inline airbyte-ci/connectors/metadata_service/orchestrator/tests/test_specs_secrets_mask.py:8]: 👍

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/specs_secrets_mask.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import dpath.util
+import yaml
+from dagster import asset
+
+GROUP_NAME = "specs_secrets_mask"
+
+# HELPERS
+
+
+def get_secrets_properties_from_registry_entry(registry_entry: dict) -> List[str]:
```
Comment (bnchrch) [inline airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/specs_secrets_mask.py:15]: Nit: a comment should be helpful here

**Review by bnchrch — COMMENTED** [review]
Looking great so far!

Replied to your questions and left a comment about merging two assets :)

File: airbyte-ci/connectors/metadata_service/lib/metadata_service/spec_cache.py (hunk 1 above)
Comment (alafanechere) [reply airbyte-ci/connectors/metadata_service/lib/metadata_service/spec_cache.py:69]: Ok, let's keep it there. We might have non-orchestrator use cases that require downloading a spec.

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/specs_secrets_mask.py (hunk 6 above)
Comment (alafanechere) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/assets/specs_secrets_mask.py:15]: Done

File: airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/__init__.py (hunk 2 above)
Comment (alafanechere) [reply airbyte-ci/connectors/metadata_service/orchestrator/orchestrator/__init__.py]: Done in 150001e62a297312c531116076e5305315fd3cac

**Comment by alafanechere** [comment]
The `specs_secret_mask.yaml` is now uploaded to GCS: https://storage.googleapis.com/dev-airbyte-cloud-connector-metadata-service/catalogs/specs_secrets_mask.yaml

**Review by bnchrch — APPROVED** [review]
Great Mask!

Have another high quality mask as a approval gif

**Pull request merged**